### PR TITLE
Fix PCA Mojo in order to play nicely with MOJO pipeline builder

### PR DESCRIPTION
**Pull request by michalkurka**
Implement a test that shows PCA can be used in MOJO Pipelines

Improvements & fixes in MOJO Import
    
    - added convenience method for loading of MOJO models
    - added an option to skip the check of white-listed algos
    - fixed an issue when the Frame MOJO is leaked (when user loads from path)
    - unsupported MOJOs don't get proper metric, they only get dummy metric object
    - fixed an issue of loading MOJOs that don't have model metadata

Fix PCA Mojo in order to play nicely with MOJO pipeline builder

### CODE REVIEW

**Comment by michalkurka** [comment]
@wendycwong the fix in PCA MOJO is minimal

@Pscheidl I can split into 2 PRs but I wanted to have the option to easily test PCA pipeline in H2O

File: h2o-algos/src/main/java/hex/generic/Generic.java [hunk 1]
```diff
@@ -81,17 +82,20 @@ public void computeImpl() {
                 final MojoReaderBackend readerBackend = MojoReaderBackendFactory.createReaderBackend(mojoBytes.openStream(_job._key), MojoReaderBackendFactory.CachingStrategy.MEMORY);
                 mojoModel = ModelMojoReader.readFrom(readerBackend, true);
                 
-                if(!ALLOWED_MOJO_ALGOS.contains(mojoModel._modelDescriptor.algoName().toLowerCase())){
-                    throw new IllegalArgumentException(String.format("Unsupported MOJO model '%s'. ", mojoModel._modelDescriptor.algoName()));
+                if(! ALLOWED_MOJO_ALGOS.contains(mojoModel._modelDescriptor.algoName().toLowerCase())) {
```
Comment (Pscheidl) [inline h2o-algos/src/main/java/hex/generic/Generic.java]: Scoring is going to work for all MOJOs. I have a different strategy on my mind - just allowing all the algos and only warning the user that model metrics are not available. Not necessarily to be solved in this PR, just a thought to work with :)

Comment (Pscheidl) [reply h2o-algos/src/main/java/hex/generic/Generic.java]: As you're using "Dummy metrics builder", we could simply make scoring work for every MOJO and only display the warning, as you wrote it.

File: h2o-algos/src/main/java/hex/generic/GenericModel.java [hunk 2]
```diff
@@ -52,34 +55,79 @@ private static MojoModel reconstructMojo(ByteVec mojoBytes) {
...
+            case AnomalyDetection:
+                return new ModelMetricsAnomaly.MetricBuilderAnomaly("");
+            default:
+                throw H2O.unimpl();
+        }
+    }
+    
+    private ModelMetrics.MetricBuilder unsupportedMetricsBuilder() {
+        if (_parms._disable_algo_check) {
+            Log.warn("Using dummny metric builder for unsupported model category: " + _output._modelCategory);
```
Comment (Pscheidl) [inline h2o-algos/src/main/java/hex/generic/GenericModel.java]: Model category 'XY' is currently unsupported. Model metrics will not be available.

File: h2o-algos/src/main/java/hex/generic/GenericModel.java [hunk 3]
```diff
@@ -52,34 +55,79 @@ private static MojoModel reconstructMojo(ByteVec mojoBytes) {
...
+        }
+
+        @Override
+        public double[] perRow(double[] ds, float[] yact, Model m) {
+            return ds;
+        }
+
+        @Override
+        public ModelMetrics makeModelMetrics(Model m, Frame f, Frame adaptedFrame, Frame preds) {
+            return null;
```
Comment (Pscheidl) [inline h2o-algos/src/main/java/hex/generic/GenericModel.java]: Why don't we return an instance of Generic metrics ?

File: h2o-algos/src/test/java/hex/pca/PCATest.java [hunk 4]
```diff
@@ -442,4 +453,59 @@ public void testPCAPredMojoPojoNumericsOnly() {
...
+      new MojoPipelineBuilder()
+              .addModel("pca", new File(pcaMojoUri))
+              .addMapping("PC1", "pca", 0)
+              .addMapping("PC2", "pca", 1)
+              .addMapping("PC3", "pca", 2)
+              .addMapping("PC4", "pca", 3)
+              .addMainModel("kmeans", new File(kmeansMojoUri))
+              .buildPipeline(pipelineFile);
+
+      GenericModel pipelineModel = Generic.importMojoModel(pipelineFile.getAbsolutePath(), true);
```
Comment (Pscheidl) [inline h2o-algos/src/test/java/hex/pca/PCATest.java]: Test the dummy metrics are in place ?

File: h2o-algos/src/main/java/hex/generic/Generic.java (hunk 1 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/generic/Generic.java]: Interesting, I think that is a good idea.

File: h2o-algos/src/main/java/hex/generic/GenericModel.java (hunk 2 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/generic/GenericModel.java]: Done, thanks

File: h2o-algos/src/test/java/hex/pca/PCATest.java (hunk 4 above)
Comment (michalkurka) [reply h2o-algos/src/test/java/hex/pca/PCATest.java]: We do not actually calculate any metrics. The metric builder is only needed for BigScore to work.

**Comment by michalkurka** [comment]
@Pscheidl thanks for the comments and the idea of allowing all mojos except for the metrics, I think that is a good idea and we should look into it

**Pull request merged**